File: src/omnia/data/structure/tuple/Septuplet.java
package omnia.data.structure.tuple;

import java.util.function.Function;

public interface Septuplet<T> extends Septuple<T, T, T, T, T, T, T>, Tuples.AtMostSeptuplet<T> {

  static <T> Septuplet<T> of(T first, T second, T third, T fourth, T fifth, T sixth, T seventh) {
    return new ImmutableSeptuplet<>(first, second, third, fourth, fifth, sixth, seventh);
  }

  @Override
  <R> Septuplet<R> map(Function<? super T, ? extends R> mapper);

  @Override
  Sextuplet<T> dropFirst();

  @Override
  Sextuplet<T> dropSecond();

  @Override
  Sextuplet<T> dropThird();

  @Override
  Sextuplet<T> dropFourth();

  @Override
  Sextuplet<T> dropFifth();

  @Override
  Sextuplet<T> dropSixth();

  @Override
  Sextuplet<T> dropSeventh();

  @Override
  Octuplet<T> concat(T object);

  @Override
  Nonuplet<T> concat(Couple<T, T> other);

  @Override
  Decuplet<T> concat(Triple<T, T, T> other);
}

File: src/omnia/data/structure/tuple/Septuplet.kt
package omnia.data.structure.tuple

import omnia.data.structure.tuple.Tuples.AtMostSeptuplet

interface Septuplet<T : Any> : Septuple<T, T, T, T, T, T, T>, AtMostSeptuplet<T> {

  override fun <R : Any> map(mapper: (T) -> R): Septuplet<R>
  override fun dropFirst(): Sextuplet<T>
  override fun dropSecond(): Sextuplet<T>
  override fun dropThird(): Sextuplet<T>
  override fun dropFourth(): Sextuplet<T>
  override fun dropFifth(): Sextuplet<T>
  override fun dropSixth(): Sextuplet<T>
  override fun dropSeventh(): Sextuplet<T>
  override fun concat(`object`: T): Octuplet<T>
  override fun concat(other: Couple<T, T>): Nonuplet<T>
  override fun concat(other: Triple<T, T, T>): Decuplet<T>

  companion object {

    fun <T : Any> of(
        first: T,
        second: T,
        third: T,
        fourth: T,
        fifth: T,
        sixth: T,
        seventh: T
    ): Septuplet<T> {
      return ImmutableSeptuplet(first, second, third, fourth, fifth, sixth, seventh)
    }
  }
}